ref(lint): add the eslint-plugin-query

TkDodo opened the pull request:
This PR adds the [eslint-plugin-query](https://tanstack.com/query/v4/docs/eslint/eslint-plugin-query) recommended set of lint rules for working with React Query. It also tries to fix most violations we have accrued.

TkDodo committed:
ref: add the eslint-plugin-query

TkDodo committed:
ref: fix @tanstack/query/no-unstable-deps violations

TkDodo committed:
ref: fix @tanstack/query/exhaustive-deps violations

TkDodo commented on static/app/components/feedback/useMutateActivity.tsx:
```diff
@@ -79,23 +79,23 @@ export default function useMutateActivity({
     (note, noteId, activity, options) => {
-      mutation.mutate([{note, noteId, activity}, 'PUT'], options);
+      mutate([{note, noteId, activity}, 'PUT'], options);
     },
-    [mutation]
+    [mutate]
```
top-level objects returned by `useQuery` and `useMutation` are _not_ referentially stable, but functions and values on them are. The linter knows this and now recommends destructuring those.

Essentially, the memoization with `useCallback` didn’t do anything before, so I’m wondering why we are having it in the first place?

TkDodo committed:
ref: store a structured queryKey

TkDodo committed:
Merge branch 'master' into tkdodo/ref/eslint-plugin-query

TkDodo commented on static/app/views/performance/newTraceDetails/traceApi/useTraceMeta.tsx:
```diff
@@ -122,7 +125,7 @@ export type TraceMetaQueryResults = {
 };
 
 export function useTraceMeta(replayTraces: ReplayTrace[]): TraceMetaQueryResults {
-  const api = useApi();
+  const fetchTraceMetaInBatches = useFetchTraceMetaInBatches();
```
this refactoring moves `useApi` into `useFetchTraceMetaInBatches` to “work around” reporting `api` as missing in the `queryKey`.

Technically, we are still potentially able to get into stale closure issues, but `useApi` doesn’t really do anything apart from storing a `new Client()` on a ref.

I’m also wondering why we don’t pass `persistInFlight: true` here, which is what `useApiQuery` does. I was under the impression that when using an api with react-query, we don’t want cancellation handled by the api client, but by react-query when opting into it.

And spinning this further, without cancellation, there isn’t anything that `useApi()` does that `const api = new Client()` wouldn’t also do?

So really, the question is: Could I also refactor these cases to just invoke `new Client()` inside the `queryFn` instead of calling `useApi` ?

JonasBa replied:
I think you could, granted you'd lose the lifecycle management from useApi. The current general recommendation is to use useApi everywhere, but I would say that if we see that it gets in the way, or if we can see opportunity to simplify things without it, we should definitely investigate them

TkDodo replied:
okay, I’m done here. I’ve moved towards `fetchDataQuery` and `useApiQuery` in as many places as I could, and disabled the linter for others.

I’d appreciate another look 🙏

TkDodo commented on static/app/views/settings/project/projectOwnership/addCodeOwnerModal.tsx:
```diff
@@ -121,11 +121,11 @@ export default function AddCodeOwnerModal({
 
-  const addFile = useCallback(() => {
+  const addFile = () => {
     if (codeownersFile) {
       mutation.mutate([{codeMappingId, raw: codeownersFile.raw}]);
     }
-  }, [codeMappingId, codeownersFile, mutation]);
+  };
```
if fixed the unstable dependency here simply by removing `useCallback`. The `addFile` function was only passed to a `Button`’s `onClick` handler, and `Button` is not a memoized component so there isn’t really a need to memoize this.

TkDodo committed:
Merge branch 'master' into tkdodo/ref/eslint-plugin-query

TkDodo committed:
Merge branch 'master' into tkdodo/ref/eslint-plugin-query

# Conflicts:
#	eslint.config.mjs
#	static/app/components/feedback/useMutateActivity.tsx
#	static/app/utils/queryClient.tsx
#	yarn.lock

TkDodo committed:
ref: run as many queries as possible with fetchDataQuery

because it uses the internal api client (so it doesn't need to go into the queryKey)

TkDodo committed:
ref: disable

TkDodo committed:
Merge branch 'master' into tkdodo/ref/eslint-plugin-query

TkDodo committed:
ref: useApiQuery in useSpanSamples

TkDodo committed:
Merge branch 'master' into tkdodo/ref/eslint-plugin-query

TkDodo committed:
chore: no idea how this worked without this mock...

TkDodo commented on static/app/views/insights/common/queries/useSpanSamples.tsx:
@JonasBa I had to re-write the mocks in tests for this change, so I wanted to sanity tests this in the UI. I can’t figure out which route I need to navigate towards to see this component though :( The nestings and abstractions in the router make this very hard to follow through. What’s the best way to figure out where this component is used in the UI ?

TkDodo committed:
Merge branch 'master' into tkdodo/ref/eslint-plugin-query

TkDodo committed:
Merge branch 'master' into tkdodo/ref/eslint-plugin-query

TkDodo committed:
Merge branch 'master' into tkdodo/ref/eslint-plugin-query
